core/src/test/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreIT.java

```diff
@@ -899,6 +903,232 @@ public void testDataFileCorruptionDuringRestore() throws Exception {
+                                                                            .setWaitForCompletion(true)
+                                                                            .setIndices(indexName)
+                                                                            .get();
+
+        assertThat(snapshotResponse.getSnapshotInfo().state(), equalTo(SnapshotState.SUCCESS));
+        assertThat(snapshotResponse.getSnapshotInfo().successfulShards(), equalTo(numShards.numPrimaries));
+        assertThat(snapshotResponse.getSnapshotInfo().failedShards(), equalTo(0));
+
+        // delete the test index
+        assertAcked(client().admin().indices().prepareDelete(indexName));
+
+        // update the test repository so that restoring files will always fail
+        assertAcked(client().admin().cluster().preparePutRepository("test-repo")
+                                              .setType("mock")
+                                              .setSettings(
+                                                    Settings.builder()
+                                                        .put("location", repositoryLocation)
+                                                        .put("random", randomAlphaOfLength(10))
+                                                        .put("max_failure_number", 10000000L)
+                                                        // No lucene corruptions, we want to test retries
+                                                        .put("use_lucene_corruption", false)
+                                                        // Restoring a file will never complete
+                                                        .put("random_data_file_io_exception_rate", 1.0)));
+
+        // attempt to restore the snapshot
+        RestoreSnapshotResponse restoreResponse = client().admin().cluster().prepareRestoreSnapshot("test-repo", "test-snap")
+                                                                            .setMasterNodeTimeout("30s")
+                                                                            .setWaitForCompletion(true)
+                                                                            .get();
+
+        // check that all shards failed during restore
+        assertThat(restoreResponse.getRestoreInfo().totalShards(), equalTo(numShards.numPrimaries));
+        assertThat(restoreResponse.getRestoreInfo().successfulShards(), equalTo(0));
+
+        // check that every shard failed because the number of allocation retries has been reached;
+        // as it can take some time and shard allocation can be throttled, we use an assertBusy()
+        assertBusy(() -> {
+            Client client = client();
+            for (int shardId = 0; shardId < numShards.numPrimaries; shardId++) {
+                ClusterAllocationExplainResponse allocationResponse = client.admin().cluster().prepareAllocationExplain()
```

Comment:
no need for allocation explain API. you can check all this directly on the cluster state that you get below. I also think that assertBusy won't be needed then.

Reply:
Right, we can use the cluster state in this case and it's even easier. But it seems that the assertBusy() is still needed as the cluster state change can take few miliseconds to propagate.